File: bindings/nodejs/lib/future.js
/*
 * future.js
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


"use strict";
var semver = require('semver');

function isFunction(f) {
	return typeof(f) == 'function';
}

function isObject(o) {
	return o === Object(o);
}

var resolvePromise = function(promise, value) {
	var called = false;
	try {
		if(promise === value)
			promise._state.reject(new TypeError('promise.then cannot be fulfilled with itself as the argument.'));

		if(isObject(value)) {
			var then = value.then;
			if(isFunction(then)) {
				then.call(value, function(res) {
					if(!called) {
						called = true;
						resolvePromise(promise, res, promise);
					}
				}, function(err) {
					if(!called) {
						called = true;
						promise._state.reject(err);
					}
				});
			}
			else
				promise._state.fulfill(value);
		}
		else
			promise._state.fulfill(value);
	}
	catch(error) {
		if(!called)
			promise._state.reject(error);
	}
};

var FuturePrototype = {
	cancel: function() {
		//cancel is not implemented for most futures
	},

	then: function(onFulfilled, onRejected) {
		var self = this;
		var future = create();
		this._state.addCallback(function(err, res) {
			var setImmediateFunc;
			if(semver.satisfies(process.version, '>=0.10.0')) {
				setImmediateFunc = setImmediate;
			}
			else {
				setImmediateFunc = process.nextTick;
			}
			setImmediateFunc(function() {
				try {
					if(self._state.rejected) {
						if(isFunction(onRejected))
							res = onRejected(err);
						else {
							future._state.reject(err);
							return;
						}
					}
					else if(isFunction(onFulfilled))
						res = onFulfilled(res);

					resolvePromise(future, res);
				}
				catch(error) {
					future._state.reject(error);
				}
			});
		});

		return future;
	},

	"catch": function(onRejected) {
		this.then(undefined, onRejected);
	}
};

FuturePrototype.__proto__ = Function.__proto__;

var FutureState = function() {
	this.callbacks = [];
	this.fulfilled = false;
	this.rejected = false;
};

FutureState.prototype.triggerCallbacks = function() {
	for(var i = 0; i < this.callbacks.length; ++i)
		this.callbacks[i](this.error, this.value);

	this.callbacks = [];
};

FutureState.prototype.addCallback = function(cb) {
	if(!this.rejected && !this.fulfilled)
		this.callbacks.push(cb);
	else
		cb(this.error, this.value);
};

FutureState.prototype.fulfill = function(value) {
	if(!this.fulfilled && !this.rejected) {
		this.fulfilled = true;
		this.value = value;
		this.triggerCallbacks();
	}
};

FutureState.prototype.reject = function(reason) {
	if(!this.fulfilled && !this.rejected) {
		this.rejected = true;
		this.error = reason;
		this.triggerCallbacks();
	}
};

var getFutureCallback = function(futureState) {
	return function(err, val) {
		if(err)
			futureState.reject(err);
		else
			futureState.fulfill(val);
	};
};

var create = function(func, cb) {
	if(cb)
		func(cb);
	else {
		// This object is used to break a reference cycle with C++ objects
		var futureState = new FutureState();

		var future = function(callback) {
			if(typeof callback === 'undefined')
				return future;

			future.then(function(val) { callback(undefined, val); }, callback);
		};

		future._state = futureState;
		future.__proto__ = FuturePrototype;

		if(func)
			func.call(future, getFutureCallback(futureState));

		return future;
	}
};

var resolve = function(value) {
	var f = create();
	f._state.fulfill(value);
	return f;
};

var reject = function(reason) {
	var f = create();
	f._state.reject(reason);
	return f;
};

var all = function(futures) {
	var future = create(function(futureCb) {
		var count = futures.length;

		if(count === 0)
			futureCb(undefined, []);

		var successCallback = function() {
			if(--count === 0)
				futureCb(undefined, futures.map(function(f) { return f._state.value; }));
		};

		for(var i = 0; i < futures.length; ++i) {
			if(futures[i] && isFunction(futures[i].then))
				futures[i].then(successCallback, futureCb);
			else
				successCallback();
		}
	});

	return future;
};

var race = function(futures) {
	var future = create(function(futureCb) {
		var successCallback = function(val) {
			futureCb(undefined, val);
		};

		for(var i = 0; i < futures.length; ++i) {
			if(futures[i] && isFunction(futures[i].then))
				futures[i].then(successCallback, futureCb);
			else {
				futureCb(undefined, futures[i]);
				break;
			}
		}
	});

	return future;
};

module.exports = {
	FuturePrototype: FuturePrototype,
	create: create,
	resolve: resolve,
	reject: reject,
	all: all,
	race: race
};

File: bindings/java/src/main/com/apple/foundationdb/async/Future.java
/*
 * Future.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.foundationdb.async;

/**
 * The typed result of an asynchronous process.
 *  A {@code Future} will be set exactly once to either a "completed" {@code T}
 *  or to an unchecked exception. A {@code Future} is a special case of a
 *  {@link PartialFuture} that cannot result in a checked exception.<br>
 * <br>
 * A {@code Future} can be waited on by blocking, or by using asynchronous callbacks.
 *
 * @param <T> the type of the eventual value of the {@code Future}
 */
public interface Future<T> extends PartialFuture<T> {
	/**
	 * Apply a function to the successful result of this operation. This version of
	 *  {@code map()}, since it accepts only {@code Function}s that cannot throw checked
	 *  exceptions, returns a {@code Future}. Errors encountered
	 *  in the execution of this operation are sent to the resulting {@code Future}.<br>
	 * <br>
	 * If the returned {@code Future} is {@link Future#cancel() cancelled} the input
	 * {@code Future} will also be cancelled.
	 *
	 * @param m the map to the conversion to be run
	 * @return a new {@code Future} that passes errors and converts output of this {@code Future}.
	 */
	public <V> Future<V> map(final Function<? super T, V> m);

	/**
	 * Applies the successful result of this operation as the input into another asynchronous
	 *  operation and returns a handle to that result.
	 *  That is, when the operation represented by this {@code Future} is complete and a result
	 *  is returned, the {@code Mapper} {@code m} is invoked to begin a new asynchronous process
	 *  using the result.  This version of
	 *  {@code map()}, since it accepts only {@code Function}s that cannot throw checked
	 *  exceptions, returns a {@code Future}.
	 *  Errors in the execution of either operation are sent to the resulting {@code Future}.<br>
	 * <br>
	 * If the returned {@code Future} is {@link Future#cancel() cancelled} the input
	 * {@code Future} will also be cancelled.
	 *
	 * @param m the map to the asynchronous process to be run
	 * @return a new {@code Future} that passes errors and converts output of this {@code Future}
	 *  to the output of another asynchronous process.
	 */
	public <V> Future<V> flatMap(final Function<? super T, Future<V>> m);

	/**
	 * Returns a {@code Future} that modifies the error behavior of this {@code Future}.
	 *  Since {@code Future} cannot be set to an {@link Exception} other than a
	 *  {@code RuntimeException}, the handling {@code Function} need not handle other,
	 *  more general, types. If an {@link Error} is the output of this process, this handler
	 *  is not invoked and that {@code Error} will be passed on to the resulting {@code Future}.
	 *
	 * @param m the process to run that maps an error into another asynchronous operation
	 *
	 * @return a newly created {@code Future} with modified error behavior
	 */
	public Future<T> rescueRuntime(Function<? super RuntimeException, Future<T>> m);

	/**
	 * Blocks until a value is set on this {@code Future} and returns it.
	 * If this {@code Future} is set to a RuntimeException or Error, throws it.
	 * A {@code Future} cannot be set to a checked exception, so there is no
	 * need to throw from this call.
	 */
	@Override
	public T get();

	/**
	 * Blocks until a value is set on this {@code Future} and returns it.
	 *  If this {@code PartialFuture} is set to an {@link Exception} or {@link Error}
	 *  this call will throw that error.
	 *
	 * @return the output value of the asynchronous process
	 *
	 * @throws InterruptedException if the blocked thread is interrupted
	 */
	@Override
	public T getInterruptibly() throws InterruptedException;

	/**
	 * Gets the error from a {@code Future} if it has been set to this state. If
	 *  this {@code Future} is not yet set, or has not been set to an error, throws
	 *  an exception. Although this method returns a {@code Throwable}, the set of
	 *  possible return types is more limited. A {@code Future} is limited to being
	 *  set to an unchecked exceptions -- that is an {@link Error} or a
	 *  {@link RuntimeException} (or one of its subclasses).
	 *
	 * @return the error output of the {@code Future}
	 *
	 * @throws IllegalStateException if this {@code Future} is not in an error state
	 */
	@Override
	public Throwable getError();
}
